File: haskell/src/Solutions/Euler058.hs
module Solutions.Euler058 where
import Utils.Monad(whenM)
import Utils.NumberTheory(primesUpTo, isPrime)
import Control.Monad
import Control.Monad.ST
import Data.STRef

result :: Int
result = runST $ do
    primes <- newSTRef (0 :: Int)
    total <- newSTRef (1 :: Int)
    step <- newSTRef (2 :: Int)
    curr <- newSTRef (1 :: Int)
    let loop = do s <- readSTRef step
                  replicateM_ 4 $ do
                      modifySTRef curr (+s)
                      modifySTRef total succ
                      whenM (isPrime <$> readSTRef curr) (modifySTRef primes succ)
                  p <- readSTRef primes
                  t <- readSTRef total
                  if 10 * p < t
                      then succ <$> readSTRef step
                      else modifySTRef step (+2) >> loop
    loop

euler058 :: IO String
euler058 = return $ show result

File: rust/src/solutions/euler058.rs
use primal::Sieve;

pub fn euler058() -> String {
    let sieve = Sieve::new(1000000000);
    let mut primes = 0;
    let mut total = 1;
    let mut step = 2;
    let mut curr = 1;
    loop {
        for _ in 0..4 {
            curr += step;
            total += 1;
            if sieve.is_prime(curr) {
                primes += 1;
            }
        }
        if 10 * primes < total {
            break (step + 1).to_string();
        } else {
            step += 2;
        }
    }
}
